### PR TITLE
[FIX] website_sale: adapt /shop page for large num of products

### PR DESCRIPTION
**Pull request by yelizariev**
The following query doesn't work well with long list of ids

`SELECT ... WHERE ... AND ..._id in <product ids>`

For such cases we can combine the query with a query for searching product ids.

Performance test
================

* 58 K product.template records
* demo attributes and categories
* Odoo v14
* postgres 12.5

/shop

```
|                         | Number of queries | Query time, sec | Remaining time, sec |
|-------------------------+-------------------+-----------------+---------------------|
| Before ProductAttribute |                   |           0.459 |                   ? |
| Before Total            |                45 |           0.538 |               0.182 |
|-------------------------+-------------------+-----------------+---------------------|
| After ProductAttribute  |                   |   0.071 + 0.030 |                     |
| After Total             |                45 |           0.189 |               0.194 |
```

/shop search with many results

```
|                              | Number of queries | Query time, sec | Remaining time, sec |
|------------------------------+-------------------+-----------------+---------------------|
| Before Category              |                   |           0.138 | ?                   |
| Before ProductAttribute      |                   |           0.530 | ?                   |
| Before Total                 |                67 |           0.821 | 0.238               |
|------------------------------+-------------------+-----------------+---------------------|
| After Category.browse(query) |                   |           0.030 | ?                   |
| After ProductAttribute       |                   |   0.207 + 0.031 | ?                   |
| After Total                  |                67 |           0.472 | 0.244               |
```

---

opw-2460065

Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by yelizariev** [comment]
@nseinlet another sql improvements to review

**Comment by nseinlet** [comment]
It seems a bit complex to "only" divide by 2 the execution time. Maybe @rco-odoo or his team have another opinion on this.

**Comment by yelizariev** [comment]
The test is done for 50K. For 500K difference should be bigger

File: addons/website_sale/controllers/main.py [hunk 1]
```diff
@@ -249,10 +249,52 @@ def shop(self, page=0, category=None, search='', ppg=False, **post):
         Product = request.env['product.template'].with_context(bin_size=True)
 
         search_product = Product.search(domain, order=self._get_search_order(post))
+        product_count = len(search_product)
```
Comment (rco-odoo) [inline addons/website_sale/controllers/main.py:252]: This is the root of all evil in this controller.
If you look carefully, you don't actually *need* `search_product`, but instead:
* `len(search_product)`, which may be given by `search_count`...
* injecting `search_product.ids` into a domain (a subquery would do the job)
* `search_product[offset: offset + ppg]`, which may be searched in a specific way

What I suggest is to use method `_search()`, which returns a `Query` object that represents the product ids:
```python
# this Query object replaces the recordset 'search_product'
product_ids = Product._search(domain)
```
You can inject `product_ids` into domains: the domain converter (expression.py) should automatically inject a subquery in the main query, and the main query is left to be optimized by PostgreSQL:
```python
search_categories = Category.search([('product_tmpl_ids', 'in', product_ids)] + website_domain).parents_and_self
```
For the actual product search, simply make it explicit:
```python
# this replaces len(search_product)
product_count = Product.search_count(domain)

# this replaces search_product[offset: offset + ppg]
products = Product.search(domain, order=self._get_search_order(post), offset=offset, limit=ppg)
```

**Review by rco-odoo — COMMENTED** [review]
This is more complicated than necessary.

**Comment by rco-odoo** [comment]
@yelizariev I also suggest to avoid the special case of the result being "not too large" at first.
Give it a try for a database with many products, it should get faster.
Then give it a try for a database with few products: add the special case only if that one is too slow.

**Comment by rco-odoo** [comment]
@yelizariev you might have a look at overrides of `_search()` in the model of products, to check whether they return a `Query` and not actual ids.

File: addons/website_sale/controllers/main.py (hunk 1 above)
Comment (yelizariev) [reply addons/website_sale/controllers/main.py:252]: I had this impression too, but the truth is that searching all records is faster than reading ``search_count`` + ``search(..., limit=..., offset=...)``. Something like 240 ms vs 190+190 ms

**Comment by yelizariev** [comment]
> Then give it a try for a database with few products: add the special case only if that one is too slow.

I did. Without special case it becomes slower on small databases

https://github.com/odoo/odoo/compare/14.0...odoo-dev:14.0-opw-2460065-website_sale-search-products--perf-IEL?expand=1

**Comment by yelizariev** [comment]
> @yelizariev you might have a look at overrides of `_search()` in the model of products, to check whether they return a `Query` and not actual ids.

only product.product is redefined without a change in result type

https://github.com/odoo/odoo/blob/0f6fdce76809af7e965f672970d74594a337f6fb/addons/product/models/product.py#L425-L429

File: addons/website_sale/controllers/main.py [hunk 2]
```diff
@@ -249,10 +249,52 @@ def shop(self, page=0, category=None, search='', ppg=False, **post):
...
+
         website_domain = request.website.website_domain()
         categs_domain = [('parent_id', '=', False)] + website_domain
         if search:
-            search_categories = Category.search([('product_tmpl_ids', 'in', search_product.ids)] + website_domain).parents_and_self
+            if product_count < request.env.cr.IN_MAX:
```
Comment (yelizariev) [inline addons/website_sale/controllers/main.py:257]: Could be a condition depending on System Parameter instead
```suggestion
            if not request.env['ir.config_parameter'].get_param('website_sale.products_optimization', False):
```

**Comment by yelizariev** [comment]
It's decided to don't merge it: it makes two versions of the same method which complicates maintaining and also doesn't have a clear criteria when it's better to use one version and not another

**Pull request closed without merging**